### PR TITLE
Add SSX React

### PR DESCRIPTION
**Pull request by skgbafa**
This PR adds a new library, `ssx-react` which wraps the ssx class for usage in react, exposing it as a Provider and a Hook.

- [x] New feature (non-breaking change which adds functionality)

How Has This Been Tested?
- [ ] Unit Tests
- [x] Integration Tests
- [x] E2E

System Configuration:
Node Version: 16.18.0
OS: MacOS 12.6
Browser Version: Chromium: 106.0.5249.119

Checklist:
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/ssx-react/Readme.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
+# SSX React
+
+SSX React provides a React Provider and Hook for using SSX in React applications. It is meant to be used in conjuction with (wagmi.sh)[https://wagmi.sh/].
```
Comment (obstropolos) [inline packages/ssx-react/Readme.md]: Make sure you're using the right link brackets here

File: examples/ssx-test-react-dapp/package.json [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  "version": "0.1.0",
+  "scripts": {
+    "dev": "next dev",
+    "build": "next build",
+    "start": "next start",
+    "lint": "next lint"
+  },
+  "dependencies": {
+    "@rainbow-me/rainbowkit": "^0.7.2",
+    "@spruceid/ssx-react": "*",
```
Comment (Juliano1612) [inline examples/ssx-test-react-dapp/package.json]: Update the @spruceid/ssx-react dependency to point to /dist and avoid the local installation error

File: examples/ssx-test-react-dapp/pages/_app.tsx [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    chain.arbitrum,
+    ...(process.env.NEXT_PUBLIC_ENABLE_TESTNETS === 'true'
+      ? [chain.goerli, chain.kovan, chain.rinkeby, chain.ropsten]
+      : []),
+  ],
+  [
+    alchemyProvider({
+      // This is Alchemy's default API key.
+      // You can get your own at https://dashboard.alchemyapi.io
+      apiKey: '_gg7wSSi0KMBsdKnGVfHDueq6xMB9EkC',
```
Comment (Juliano1612) [inline examples/ssx-test-react-dapp/pages/_app.tsx]: Using this API key as default triggers a network error. Why not point to an environment variable?

File: examples/ssx-test-react-dapp/pages/index.tsx [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+        <h1 className={styles.title}>
+          Welcome to <a href="">RainbowKit</a> + <a href="">wagmi</a> +{' '}
+          <a href="https://nextjs.org">Next.js!</a>
+        </h1>
+
+        <p className={styles.description}>
+          Sign-in with Ethereum powered by SSX
+          <br/>
+          {/* <button onClick={() => {console.log(ssx)}}>Sign Message</button> */}
```
Comment (Juliano1612) [inline examples/ssx-test-react-dapp/pages/index.tsx]: Remove the comment

**Comment by Juliano1612** [comment]
Make sure to add the .npmignore file in packages/ssx-react

File: examples/ssx-test-react-dapp/pages/_app.tsx (hunk 3 above)
Comment (samgbafa) [reply examples/ssx-test-react-dapp/pages/_app.tsx]: Done

File: examples/ssx-test-react-dapp/package.json (hunk 2 above)
Comment (samgbafa) [reply examples/ssx-test-react-dapp/package.json]: Done

File: examples/ssx-test-react-dapp/pages/index.tsx (hunk 4 above)
Comment (samgbafa) [reply examples/ssx-test-react-dapp/pages/index.tsx]: Done

File: packages/ssx-react/Readme.md (hunk 1 above)
Comment (samgbafa) [reply packages/ssx-react/Readme.md]: Done. Updated to match gitbook

File: examples/ssx-test-react-dapp/.env.example [hunk 5]
```diff
@@ -0,0 +1 @@
+ALCHEMY_API_KEY=
```
Comment (Juliano1612) [inline examples/ssx-test-react-dapp/.env.example]:
```suggestion
NEXT_PUBLIC_ENABLE_TESTNETS=
NEXT_PUBLIC_ALCHEMY_API_KEY=
```

File: examples/ssx-test-react-dapp/pages/_app.tsx [hunk 6]
```diff
@@ -0,0 +1,64 @@
+import '../styles/globals.css';
+import '@rainbow-me/rainbowkit/styles.css';
+import type { AppProps } from 'next/app';
+import { RainbowKitProvider, getDefaultWallets } from '@rainbow-me/rainbowkit';
+import { chain, configureChains, createClient, WagmiConfig } from 'wagmi';
+import { alchemyProvider } from 'wagmi/providers/alchemy';
+import { publicProvider } from 'wagmi/providers/public';
+import { SSXProvider } from '@spruceid/ssx-react';
+
+if (!process.env.ALCHEMY_API_KEY) {
```
Comment (Juliano1612) [inline examples/ssx-test-react-dapp/pages/_app.tsx]:
```suggestion
if (!process.env.NEXT_PUBLIC_ALCHEMY_API_KEY) {
```

File: examples/ssx-test-react-dapp/pages/_app.tsx [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+  throw new Error('Missing ALCHEMY_API_KEY environment variable. Add to .env.local');
```
Comment (Juliano1612) [inline examples/ssx-test-react-dapp/pages/_app.tsx]:
```suggestion
  throw new Error('Missing NEXT_PUBLIC_ALCHEMY_API_KEY environment variable. Add to .env.local');
```

File: examples/ssx-test-react-dapp/pages/_app.tsx [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    chain.arbitrum,
+    ...(process.env.NEXT_PUBLIC_ENABLE_TESTNETS === 'true'
+      ? [chain.goerli, chain.kovan, chain.rinkeby, chain.ropsten]
+      : []),
+  ],
+  [
+    alchemyProvider({
+      // This is Alchemy's default API key.
+      // You can get your own at https://dashboard.alchemyapi.io
+      apiKey: process.env.ALCHEMY_API_KEY,
```
Comment (Juliano1612) [inline examples/ssx-test-react-dapp/pages/_app.tsx]:
```suggestion
      apiKey: process.env.NEXT_PUBLIC_ALCHEMY_API_KEY,
```

**Review by Juliano1612 — APPROVED** [review]
looks good to me!

File: packages/ssx-sdk/src/core.ts [hunk 9]
```diff
@@ -127,11 +132,11 @@ export class SSXConnected {
     }
   }
 
-  public async ssxServerNonce(params: Record<string, any>): Promise<{ nonce: string }> {
+  public async ssxServerNonce(params: Record<string, any>): Promise<string> {
```
Comment (161chihuahuas) [inline packages/ssx-sdk/src/core.ts:135]: This function returns `undefined` if `this.api` is falsy - is that the intended behavior?

Comment (161chihuahuas) [reply packages/ssx-sdk/src/core.ts:135]: Also see my note in the general review about `try...catch`!

**Review by 161chihuahuas — APPROVED** [review]
Overall looks good! One inline comment. General feedback (hopefully helpful!): I've noticed throughout the SSX codebase there is liberal use of `try...catch`, which is totally fine except that any code inside a `try...catch` is not optimized by the JIT. Replacing the contents of a `try` block with a function call is one way to optimize this. For example:

```js
// will be optimized by jit
function doAThingThatMightThrow() {
    // ...
}
function handleError(e) {
  // ...
}
try {
  doAThingThatMightThrow(); // calling jit optimized fn defined outside of block
} catch (e) {
  handleError(e); // calling jit optimized fn outside of block
}

// NOT optimized by JIT
try {
  const x = true;
  if (x) {
    // ...
  }
} catch (e) {
  if (e.type === 'y') {
    console.error(e.message);
  }
}
```

Just a side note but not a blocker for release. In the future try to consolidate the contents of try blocks into function defined outside of the block and call from within to take advantage of the JIT.

File: packages/ssx-sdk/src/core.ts (hunk 9 above)
Comment (samgbafa) [reply packages/ssx-sdk/src/core.ts:135]: Yes, [this is intended](https://github.com/spruceid/ssx/pull/10/files/fe82e7fed678495aab6986ae7da4d4bf7d27308e#diff-92023938c9c08f9a708ddadca984236b869fbfe8b4622000e0fd4d4e3f859220R194)!

**Pull request merged**